release-process.md

```diff
@@ -1,54 +1,131 @@
-Verify that ``master/main`` contains the commits intended for the release.  If
-there isn't already an issue in GitHub for the release, create one.
-### 2. Create a ``release-candidate`` branch and hard reset it to ``master/main``
-The ``release-candidate`` branch is hard reset to ``master/main`` since for the
-purposes of the release process, its state on the remote is irrelevant.
-### 4. Generate release notes
-All of our projects contain a file for release notes in the project
-root directory.  This file is named ``RELEASE.rst``.  For each release it is
-updated with a section titled "Release <version-number>" containing a list
-of the commits in the release.  
-### 5. Update version numbers
-Project version numbers are typically stored in ``settings.py`` and/or
-``setup.py`` depending on the application's architecture and purpose.
-### 6. Commit updates and push ``release-candidate`` branch
-### 7. Generate release notes with checkboxes
-### 8. Open PR to merge ``release-candidate`` branch into ``release`` branch
-### 9. Merge PR once developers verify their commits (manual step)
-### 10. Tag the release
-### 11. Merge the ``release`` branch into the ``master/main`` branch and push
-### 12. Send email notifications (manual step)
-
-
-## Release Version Numbers
-Our releases are identified by a version number that follows the
+  - Code starts out in a pull request which is reviewed by other team members (see [Code Review](code-review.md)).
+  - Then the pull request is merged to the `main` branch (or `master` for legacy projects).
+
+### Process
+
+```mermaid
+sequenceDiagram
+  actor eng as Software Engineers
+  participant branchmain as main/master
+
+  participant branchrc as release-candidate
+  participant envrc as Release Candidate
+
+  participant branchrelease as release
+  participant envprod as Production
+
+  loop
+    eng->>branchmain: Merge pull requests
+  end
+
+  Note over eng: Requests a release candidate
+
+  critical New Release Candidate
+    branchmain->>branchrc: Latest code merged
+    branchrc->>branchrc: Release notes and PR generated
+    branchrc->>envrc: Deploy to RC
+    eng->>envrc: Engineer verifies their commit(s)
+    note over eng: Checks off boxes on the PR
+  end
+
+  critical Production Release
+    eng->>branchrc: Engineer coordinated release approval
+    branchrc->>branchmain: Merged back to main
+    branchmain->>branchrelease: Main branch merged to release
+    branchrelease->>envprod: Deploy to Production
+  end
+
+
+```
+
+#### CI
+
+  - Deployments to the CI environment are only applicable to web applications.
+  - After tests pass on the `master`/`main` branch this triggers a deployment to a CI (continuous integration) server for web application projects.
+
+#### Release candidate
+  
+  - On a daily basis each project is checked on an automated schedule for new changes:
+    - If there are changes a release candidate is created in the `release-candidate` branch.
```

Comment:
I'm not quite sure I follow, but the main reason for the different branches is the separate environments deploy from each branch. Another reason for different branches is for hotfixes.

Reply:
I just wanted to play devil's advocate to make sure that we're not just using separate branches because "we've always done it that way", and that we have some reasoning behind that. I think that right now we are locked into that pattern because of how Heroku handles automatic deploys, but in the absence of that constraint would we still use the separate branches? Again, I don't have a strong opinion, just making sure that we're being deliberate about our approach.

Reply:
I'm not sure if there's another way to be able to cherry pick in changes and progress them through the environments without branches, do you know of any?